provider/providers.go

```diff
@@ -191,3 +188,54 @@ func NewMicrosoft(p Params) Oauth2Handler {
 		},
 	})
 }
+
+// NewPatreon makes patreon oauth2 provider
+func NewPatreon(p Params) Oauth2Handler {
+	type uinfo struct {
+		Data struct {
+			Attributes struct {
+				FullName string `json:"full_name"`
+				ImageURL string `json:"image_url"`
+			} `json:"attributes"`
+			ID            string `json:"id"`
+			Relationships struct {
+				Pledges struct {
+					Data []struct {
+						ID   string `json:"id"`
+						Type string `json:"type"`
+					} `json:"data"`
+				} `json:"pledges"`
+			} `json:"relationships"`
+		} `json:"data"`
+	}
+
+	return initOauth2Handler(p, Oauth2Handler{
+		name: "patreon",
+		// see https://docs.patreon.com/?shell#oauth
+		endpoint: oauth2.Endpoint{
+			AuthURL:   "https://www.patreon.com/oauth2/authorize",
+			TokenURL:  "https://api.patreon.com/oauth2/token",
+			AuthStyle: oauth2.AuthStyleInParams,
+		},
+		scopes:  []string{},
+		infoURL: "https://www.patreon.com/api/oauth2/api/current_user",
+
+		mapUser: func(data UserData, bdata []byte) token.User {
+			userInfo := token.User{}
+
+			uinfoJSON := uinfo{}
+			if err := json.Unmarshal(bdata, &uinfoJSON); err == nil {
+				userInfo.ID = "patreon_" + token.HashID(sha1.New(), userInfo.ID)
+				userInfo.Name = uinfoJSON.Data.Attributes.FullName
+				userInfo.Picture = uinfoJSON.Data.Attributes.ImageURL
+
+				// check if the user is your subscriber
+				if len(uinfoJSON.Data.Relationships.Pledges.Data) > 0 {
```

Comment:
I'm not sure how this supposed to work. It looks like the check is not checking for "my subscriber" but rather making sure the user subscribed to something, which may or may not be the one we care about.

Reply:
This decision was made based on this information at Patreon API documentation.

> This optional parameter will default to users pledges-to-me my-campaign, which fetches user profile information, pledges to your creator, and your creator info. It will be displayed to the user in human-friendly terms when signing in with Patreon. If your client requires the ability to ask for pledges or campaign data of other users (not just your own campaign), please email platform@patreon.com, and we'll do our best to get back to you shortly.

So that users should be "your" subscriber.

Reply:
unless I'm missing something, I don't even see it passes my campaign id in any form. How is this even supposed to know which campaign is mine?

Reply:
I may only assume that they have some kind of mapping between the provided access token and client_id.
I can add another subscription to my account and double check this assumption.

Reply:
I guess the source of my confusion is the lack of any mention of campaign id in this instruction. I'm not that familiar with Patreon, but as far as I recall under the same account user may have multiple campaigns/subscriptions?

#### Patreon Auth Provider ####
1.	Create a new Patreon client https://www.patreon.com/portal/registration/register-clients
1.  Fill **"App Name"**, **"Description"**, **"App Category"** and **"Author"** for your site
1.  Under **"Redirect URIs"** enter the correct url constructed as domain + `/auth/patreon/callback`. ie `https://example.mysite.com/auth/patreon/callback`
1.  Take note of the **Client ID** and **Client Secret**

Reply:
That confused me too. But as I've mentioned above campaign_id is not needed. I guess that they do some mapping between provided access_token and client_id that was provided during the first OAuth step. And based on this relation they return only pledges that related to your account.

Reply:
> I can add another subscription to my account and double check this assumption.

this would be very nice to make sure it works the way we expect

Reply:
Ok. So I I've tested it and received only 1 pledge for my account.

`            "relationships": {
                "campaign": {
                    "data": {
                        "id": "7381928",
                        "type": "campaign"
                    },
                    "links": {
                        "related": "https://www.patreon.com/api/campaigns/7381928"
                    }
                `
If you want I can share client ID / secret for my test account with you.